usl_pipeline/usl_lib/usl_lib/transformers/raster_transformers.py

```diff
@@ -0,0 +1,73 @@
+from typing import Tuple
+
+import numpy
+import numpy.typing as npt
+from shapely import geometry  # import Point, Polygon
+import geopandas as gpd
+
+from usl_lib.shared import geo_data
+
+
+def fill_in_soil_classes_missing_values_from_nearest_polygons(
+    elevation_header: geo_data.ElevationHeader,
+    green_areas_raster: npt.NDArray[numpy.float64],
+    soil_classes_raster: npt.NDArray[numpy.float64],
+    soil_classes_polygon_masks: list[Tuple[geometry.Polygon, int]],
+) -> list[Tuple[geometry.Polygon, int]]:
+    """Generates 1-cell polygons with nearest-neighbor soil classes for missing cells.
+
+    Args:
+        elevation_header: Elevation header providing coordinate system and the study
+            area region.
+        green_areas_raster: Raster with 0/1 masks indicating the presence of green area
+            for each raster cell.
+        soil_classes_raster: Raster with integer soil classes assigned to each raster
+            cell (0 means missing soil class).
+        soil_classes_polygon_masks: List of tuples containing polygons with associated
+            soil classes.
+
+    Returns:
+        The list of tuples containing polygons with associated soil classes for the
+        cells in a raster with present green area but missing soil class.
+    """
+    if len(soil_classes_polygon_masks) == 0:
+        return []
+
+    masks = [polygon_mask[1] for polygon_mask in soil_classes_polygon_masks]
+    d = {"geometry": [polygon_mask[0] for polygon_mask in soil_classes_polygon_masks]}
+    gdf = gpd.GeoDataFrame(d)
+
+    # Prepare coordinates left-upper corner of the study area
+    cell_size = elevation_header.cell_size
+    min_x = elevation_header.x_ll_corner
+    max_y = elevation_header.y_ll_corner + elevation_header.row_count * cell_size
+
+    # Make common 3-dimensional array where both 2d arrays are 2 layers and 3-rd
+    # dimension is just 2-element array [<green-area-cell>, <soil-class-cell>]
+    # corresponding to the same place in the raster.
+    stacked_raster = numpy.dstack((green_areas_raster, soil_classes_raster))
+
+    # Reduce 3-rd dimension to an indicator that green area is present but soil class is
+    # 0 (missing value).
+    missing_soil_classes = numpy.apply_along_axis(
+        lambda gs: gs[0] == 1 and gs[1] == 0, 2, stacked_raster
+    )
```

Comment:
oh my! I spent half a day yesterday figuring out how to correspond elements for the same positions between 2 nd-arrays. Your 1-liner just solved it!

Reply:
this is the magic of numpy!
python just provides a DSL for what are ultimately C (and even fortran) functions which make use of vector hardware. It is a very nice DSL for vector operations!

Reply:
so the secret here is that 
`green_areas_raster == 1`
and
`soil_classes_raster == 0`
both return mask arrays which you can 'and' together with the `&` operator

no need to actually join them together since they're the same dimensions with the same indices already!